assets/javascripts/discourse/components/d-templates/modal/form.js

```diff
@@ -0,0 +1,21 @@
+import Component from "@glimmer/component";
+import { service } from "@ember/service";
+import { bind } from "discourse-common/utils/decorators";
+
+export default class Form extends Component {
+  @service appEvents;
+
+  @bind
+  bindEvents() {
+    if (this.args.closeModal) {
+      this.appEvents.on("page:changed", this, this.args.closeModal);
+    }
+  }
+
+  @bind
+  unbindEvents() {
+    if (this.args.closeModal) {
+      this.appEvents.off("page:changed", this, this.args.closeModal);
+    }
+  }
```

Comment:
You can rely on the default lifecycle methods (then there's no need for the template changes)

```suggestion
  constructor() {
    super(...arguments);
    if (this.args.closeModal) {
      this.appEvents.on("page:changed", this, this.args.closeModal);
    }
  }

  willDestroy() {
    super.willDestroy(...arguments);
    if (this.args.closeModal) {
      this.appEvents.off("page:changed", this, this.args.closeModal);
    }
  }
```